### PR TITLE
Make pin rm not try to find pinned ancestors by default

### PR DESCRIPTION
**Pull request by Voker57**
Also make a switch to pin rm to allow old behavior.
Trying to look up pins which interfere with requested unpin
can be an extremely costly operation and typically is not
required by user.

### CODE REVIEW

**Comment by whyrusleeping** [comment]
@Voker57 Is this for the case where youre unpinning something 'recursively' and its pinned directly? I'm unsure the exact usecase here.

**Comment by Voker57** [comment]
@whyrusleeping it's for the case when you're trying to unpin something which is not directly pinned: currently, in that case ipfs checks all the pins to see if they affect the requested nodes, and that can take huge amounts of time in case of big pinned items.

File: test/sharness/t0080-repo.sh [hunk 1]
```diff
@@ -98,7 +104,7 @@ test_expect_success "remove recursive pin, add direct" '
 
 test_expect_success "remove direct pin" '
 	echo "unpinned $HASH" >expected6 &&
-	ipfs pin rm "$HASH" >actual6 &&
+	ipfs pin rm -r=false "$HASH" >actual6 &&
```
Comment (whyrusleeping) [inline test/sharness/t0080-repo.sh]: This behaviour shouldnt change though, `ipfs pin rm` should remove either a recursive or a direct pin

Comment (Voker57) [inline test/sharness/t0080-repo.sh]: But then there is no way to remove only recursive pins. -r will remove direct pins as well. Is that intended way for -r to behave?

**Comment by Voker57** [comment]
@whyrusleeping IMHO default behavior should be "do not explain", since it involves potentially extremely costly operation and chances that user needs resulting information are quite low. Trying to unpin a certain block/file? Then they should know ahead to run pin rm -e to determine pin source. My target case is unpinning already unpinned hash, and I think this is more likely to happen.

File: test/sharness/t0080-repo.sh (hunk 1 above)
Comment (whyrusleeping) [inline test/sharness/t0080-repo.sh]: Hrm... That was the behavior we initially intended to have. The idea was to mimic the unix `rm` command:

```sh
rm file #works
rm dir #doesnt work
rm -r file #works
rm -r dir #works
```

This does bring up the equivalent of having an `rmdir` analogue though.

What do you think @jbenet @Kubuxu @lgierth  ?

**Comment by whyrusleeping** [comment]
@Voker57 Okay, you've convinced me that having do not explain be the default is the right thing to do. That feels more natural to me now

**Comment by whyrusleeping** [comment]
(bumping to 0.4.7 as getting feedback on changing api behaviour will take a few days)

File: test/sharness/t0080-repo.sh (hunk 1 above)
Comment (Voker57) [inline test/sharness/t0080-repo.sh]: It's not mimicing rm -r anyway, since it will not delete included pins.

**Comment by whyrusleeping** [comment]
@Kubuxu @lgierth @kevina could I get you guys to review this?

**Comment by Kubuxu** [comment]
Regarding code LGTM, regarding UI, I am not sure but this probably applies to current UI too.

File: test/sharness/t0080-repo.sh (hunk 1 above)
Comment (kevina) [inline test/sharness/t0080-repo.sh]: I agree that `rm` is a bad model.  The type of a pin does not say anything about the type of a file. In addition I image most pins are recursive and direct pins are only used in special cases.

**Comment by kevina** [comment]
This very closely relates to #3796.  Give me a few days to try this p.r. out and verify the current and changed behavior of of `ipfs pin rm`.

File: test/sharness/t0080-repo.sh (hunk 1 above)
Comment (kevina) [inline test/sharness/t0080-repo.sh]: I partly take back what I said.  I left another comment.

**Comment by kevina** [comment]
i don't like the change in behavior  `pin --recursive=true` should remove both recursive a direct pins, `pin --recursive=false` should only remove direct pins.

**Comment by Voker57** [comment]
How to remove only recursive pins them? Technically they are completely unrelated, imo interface implying they are is misleading.

**Comment by kevina** [comment]
I'm not sure, but, in my option there should be a way to remove a pin regardless of it's type.  We should not change the default behavior in any case.

Perhaps the recursive option should be removed, and replaced with a `--type` option that defaults to any, this will make it more explicit what is going on.  Or perhaps we can keep the recursive option as is and in addition add a `--type` option.

@whyrusleeping (and others) opinions?

**Comment by Voker57** [comment]
It's not related to PR anyways, i just changed it because I thought current behavior is not intentional and test was not clear on that. I'll remove the change if -r is indeed supposed to remove both pins.

**Comment by kevina** [comment]
Without explain enabled doing a `ipfs pin rm` on a direct pin will now return `Error: not pinned` which is misleading at best.

**Comment by Voker57** [comment]
I've removed changing pin rm -r behavior from PR.

File: pin/pin.go [hunk 2]
```diff
@@ -84,7 +84,7 @@ type Pinner interface {
 	IsPinned(*cid.Cid) (string, bool, error)
 	IsPinnedWithType(*cid.Cid, PinMode) (string, bool, error)
 	Pin(context.Context, node.Node, bool) error
-	Unpin(context.Context, *cid.Cid, bool) error
+	Unpin(context.Context, *cid.Cid, bool, bool) error
```
Comment (kevina) [inline pin/pin.go]: Please add parameter to Pin and Unpin so it is clear what the bool parameters are far.

Comment (Voker57) [inline pin/pin.go]: named.

**Comment by Voker57** [comment]
Appears to be already changed during some overhaul.

**Comment by Stebalien** [comment]
This was fixed in #6311. I had no idea there was an existing PR.

**Pull request closed without merging**